README.md
# ieprader

FIXME: my new application.

## Installation

Download from https://github.com/pradesigner/ieprader

## Usage

FIXME: explanation

Run the project directly, via `:exec-fn`:

    $ clojure -X:run-x
    Hello, Clojure!

Run the project, overriding the name to be greeted:

    $ clojure -X:run-x :name '"Someone"'
    Hello, Someone!

Run the project directly, via `:main-opts` (`-m pradesigner.ieprader`):

    $ clojure -M:run-m
    Hello, World!

Run the project, overriding the name to be greeted:

    $ clojure -M:run-m Via-Main
    Hello, Via-Main!

Run the project's tests (they'll fail until you edit them):

    $ clojure -M:test:runner

Build an uberjar:

    $ clojure -X:uberjar

Run that uberjar:

    $ java -jar ieprader.jar

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 Pradmin

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-06-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-06-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/pradesigner/ieprader/compare/0.1.1...HEAD
[0.1.1]: https://github.com/pradesigner/ieprader/compare/0.1.0...0.1.1


test/pradesigner/ieprader_test.clj
(ns pradesigner.ieprader-test
  (:require [clojure.test :refer :all]
            [pradesigner.ieprader :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/pradesigner/nanark.clj
"
identifies stocks common to ark and nanalyze
order is in ark preference
"

(ns user
  (:require [tablecloth.api :as tc]
            [clojure.string :as s]))

(def ark (-> (tc/dataset "resources/ark.txt"
                         {:dataset-name "ark"
                          :header-row? false
                          :column-blacklist ["column-1"]})
             (tc/rename-columns {"column-0" :ticker})))
(tc/row-count ark)
;; => 161

(def nan (-> (tc/dataset "resources/nan.csv"
                         {:dataset-name "nan"
                          :header-row? false})
             (tc/select-columns ["column-9"])
             (tc/drop-rows [0])
             (tc/rename-columns {"column-9" :ticker})))

(tc/row-count nan)
;; => 326

(def nanark (-> (tc/dataset)
                (tc/add-column :nan (tc/column nan :ticker))
                (tc/add-column :ark (tc/column ark :ticker) :na)
                (tc/add-column :common ((tc/inner-join ark nan :ticker)
                                        :ticker)
                               :na)))

;; a convoluted way to sort
(-> (tc/dataset)
        (tc/add-column :arknan (tc/column (-> (tc/inner-join ark nan :ticker)
                                          (tc/group-by :ticker)
                                          (tc/ungroup {:order? true}))
                                          :ticker))
        (tc/add-column :nanark (tc/column (-> (tc/inner-join nan ark :ticker)
                                              (tc/group-by :ticker)
                                              (tc/ungroup {:order? true}))
                                          :ticker)))

;; sort
(def nanarks (-> (tc/semi-join ark nan :ticker)
                 (tc/order-by :ticker)))

(seq (nanarks :ticker))

(binding [tech.v3.dataset.print/*default-table-row-print-length* 66]
  nanarks) ;still only 25

(binding [tech.v3.dataset.print/*default-table-row-print-length* 66]
  (println nanarks)) ;prints all

(seq (nanarks :ticker))

(defn set-data-print-length! [n]
   (alter-var-root #'tech.v3.dataset.print/*default-table-row-print-length* (fn [_] n)))
(set-data-print-length! 66) ;prints all




;; * joins

(def x (tc/dataset {:a [1 2 3 4]}))
(def y (tc/dataset {:a [5 3 4 6]}))
(tc/semi-join x y :a)
;; ** left-right-full
(tc/left-join x y :a)
;; => left-outer-join [4 2]:

| :a | :right.a |
|---:|---------:|
|  3 |        3 |
|  4 |        4 |
|  1 |          |
|  2 |          |

(tc/right-join x y :a)
;; => right-outer-join [4 2]:

| :a | :right.a |
|---:|---------:|
|  3 |        3 |
|  4 |        4 |
|    |        5 |
|    |        6 |

(tc/full-join x y :a)
;; => full-join [6 2]:

| :a | :right.a |
|---:|---------:|
|  3 |        3 |
|  4 |        4 |
|  1 |          |
|  2 |          |
|    |        5 |
|    |        6 |


;; ** intersections
(tc/inner-join x y :a)
;; => inner-join [2 1]:

| :a |
|---:|
|  3 |
|  4 |

(tc/semi-join x y :a);; => semi-join [2 1]:

| :a |
|---:|
|  3 |
|  4 |

(tc/intersect (tc/select-columns x :a)
              (tc/select-columns y :a))
;; => intersection [2 1]:

| :a |
|---:|
|  3 |
|  4 |


;; ** anti
(tc/anti-join x y :a)
;; => anti-join [2 1]:

| :a |
|---:|
|  1 |
|  2 |

(tc/anti-join y x :a)
;; => anti-join [2 1]:

| :a |
|---:|
|  5 |
|  6 |

(tc/difference (tc/select-columns x :a)
               (tc/select-columns y :a))
;; => difference [2 1]:

| :a |
|---:|
|  1 |
|  2 |

(tc/difference (tc/select-columns y :a)
               (tc/select-columns x :a))
;; => difference [2 1]:

| :a |
|---:|
|  5 |
|  6 |


;; ** building a new dataset with inner join
(-> (tc/dataset)
    (tc/add-column :c (tc/column x :a))
    (tc/add-column :d (tc/column y :a))
    (tc/add-column :inner ((tc/inner-join x y :a) ;get the inner ds
                           :a) ;turn into seq
                   :na) ;deal with shortness
    )


nanarks


src/pradesigner/ieprader.clj
"
displays and summarizes results from ieprader
"

(ns pradesigner.ieprader
  (:require [tablecloth.api :as tc]
            [clojure.java.io :refer [file]]
            [clojure.string :as s]
            [clojure.java.shell :refer [sh]])
  (:gen-class))


(def fpaths
  "defines the file paths for AccountHoldings"
  (let [base "/home/pradmin/dd/"
        fns (filter #(re-find #"^Account" %)
                    (seq (.list (file base))))]
    (map #(str base %) fns)))

(defn remove-commas
  "removes commas from numbers for conversion purposes"
  [ds col]
  (map #(s/replace % #"," "") (ds col)))

(defn currency-format
  "takes to 2 decimal places"
  [ds col]
  (map #(format "%.2f" %) (ds col)))

(defn mk-ds
  "makes dataset from filepath"
  [fp]
  (tc/dataset fp
              {:header-row? false
               :key-fn keyword}))

(defn get-acct
  "gets the account number from the dataset"
  [ds]
  (->> (first (ds :column-0))
       (re-find #"(^.+) -")
       (second)))

(defn prep-columns
  "remove extraneous columns and rename existing ones"
  [ds]
  (-> ds
      (tc/rename-columns {:column-2 :ticker
                          :column-5 :qty
                          :column-6 :marketvalue
                          :column-7 :bookvalue
                          :column-8 :diff
                          :column-9 :%diff
                          :column-10 :%port})
      (tc/select-rows (comp #(= % "Equities") :column-0))
      (tc/drop-columns (for [n [0 1 3 4 11 12 13 14 15]]
                         (keyword (str "column-" n))))))

(defn str->num
  "convert from string to numbers"
  [ds]
  (-> ds
      (tc/add-or-replace-columns {:marketvalue (remove-commas ds :marketvalue)
                                  :bookvalue (remove-commas ds :bookvalue)
                                  :diff (remove-commas ds :diff)})
      (tc/convert-types {:qty :int16
                         :marketvalue :float64
                         :bookvalue :float64
                         :diff :float64
                         :%diff :float64
                         :%port :float64})))

(defn sumfn
  "sums items using reduce"
  [itm]
  (reduce + itm))

(defn prep-one-ds
  "prepares items from a single ds given filepath"
  [fp cols-to-sum]
  (let [ds-org (mk-ds fp)
        acct-info (get-acct ds-org)
        preped-ds (-> ds-org
                      (prep-columns)
                      (str->num))
        sums-of-cols (tc/aggregate-columns preped-ds
                                           cols-to-sum
                                           sumfn)
        summed-ds (tc/bind preped-ds sums-of-cols)]
    [acct-info sums-of-cols summed-ds]))

(defn prep-all-ds
  "puts all info into a sequence for further processing"
  [fps cols-to-sum]
  (let [all-data (map #(prep-one-ds % cols-to-sum) fps)
        col-sums (apply tc/bind (map second all-data))
        tot-col-sums (tc/aggregate-columns col-sums
                                           cols-to-sum
                                           sumfn)
        joined-data (for [e all-data]
                      (str "\n" (first e) "\n"
                           (s/replace (last e)
                                      #"null \[.+\]:\n\n"
                                      "")))
        final (str (s/replace tot-col-sums
                              #"_unnamed .+:\n\n"
                              "grand totals:\n")
                   "\n\n"
                   (apply str joined-data))]
    final))


(defn -main
  "runs get-group-tots"
  [& args]
  (spit "/home/pradmin/clj/ieprader/out.txt" (prep-all-ds fpaths [:marketvalue :bookvalue :diff]))
  (sh "emacs" "/home/pradmin/clj/ieprader/out.txt"))


src/pradesigner/iepraderv1.clj
"
displays and summarizes results from ieprader

TODO
pull items for accumulations from each of the accounts - by merging them?
"

(ns pradesigner.ieprader
  (:require [tablecloth.api :as tc]
            [clojure.java.io :refer [file]])
  (:gen-class))


(def fpaths
  "defines the file paths for AccountHoldings"
  (let [base "/home/pradmin/dd/"
        fns (filter #(re-find #"^Account" %)
                    (seq (.list (file base))))]
    (map #(str base %) fns)))

(defn remove-commas
  "removes commas from numbers for conversion purposes"
  [ds col]
  (map #(clojure.string/replace % #"," "") (ds col)))

(defn currency-format
  "takes to 2 decimal places"
  [ds col]
  (map #(format "%.2f" %) (ds col)))

(defn mk-ds
  "makes dataset from filepath"
  [fp]
  (tc/dataset fp
              {:header-row? false
               :key-fn keyword}))

(defn get-acct
  "gets the account number from the dataset"
  [ds]
  (->> (first (ds :column-0))
       (re-find #"(^.+) -")
       (second)))

(defn prep-columns
  "remove extraneous columns and rename existing ones"
  [ds]
  (-> ds
      (tc/rename-columns {:column-2 :ticker
                          :column-5 :qty
                          :column-6 :marketvalue
                          :column-7 :bookvalue
                          :column-8 :diff
                          :column-9 :%diff
                          :column-10 :%port})
      (tc/select-rows (comp #(= % "Equities") :column-0))
      (tc/drop-columns (for [n [0 1 3 4 11 12 13 14 15]]
                         (keyword (str "column-" n))))))

(defn str->num
  "convert from string to numbers"
  [ds]
  (-> ds
      (tc/add-or-replace-columns {:marketvalue (remove-commas ds :marketvalue)
                                  :bookvalue (remove-commas ds :bookvalue)
                                  :diff (remove-commas ds :diff)})
      (tc/convert-types {:qty :int16
                         :marketvalue :float64
                         :bookvalue :float64
                         :diff :float64
                         :%diff :float64
                         :%port :float64})))

(defn concat-totals
  "create totals from columns :qty :marketvalue :bookvalue :diff and concat to ds"
  [ds]
  (tc/concat ds
             (-> ds
                 (tc/aggregate-columns [:qty :marketvalue :bookvalue :diff] #(reduce + %))
                 (tc/add-columns {:%diff nil
                                  :%port nil
                                  :ticker "Totals"}))))

(defn format->currency
  "format specific columns to currency (note using fn to save repetition)"
  [ds]
  (-> ds
      (tc/add-or-replace-columns (apply merge
                                        (map (fn [k] {k (currency-format ds k)})
                                             [:marketvalue :bookvalue])))
      #_(tc/add-or-replace-columns {:marketvalue (currency-format ds5 :marketvalue)
                                    :bookvalue (currency-format ds5 :bookvalue)})))

(defn remove-ds-header
  "removes the ds header which shows filename and shape"
  [ds]
  (-> ds
      (tc/dataset->str)
      (clojure.string/replace #".+\n\n" "")))

(defn pr-ds
  "prepares ds given filepath"
  [fp]
  (let [ds (mk-ds fp)
        acct (get-acct ds)
        ds-str (-> ds
                   (prep-columns)
                   (str->num)
                   (concat-totals)
                   (format->currency)
                   (remove-ds-header))]
    (str acct "\n\n" ds-str "\n\n\n\n")))

(spit "resources/ieholdings.txt" (apply str (map pr-ds fpaths)))

;; * zz
(defn pr-ds
  "prepares dataset"
  [fp]
  (let [ds1 (mk-ds fp)
        acct (get-acct ds1)
        
        ;; remove extraneous columns and rename existing ones 
        ds2 (-> ds1
                (tc/rename-columns {:column-2 :ticker
                                    :column-5 :qty
                                    :column-6 :marketvalue
                                    :column-7 :bookvalue
                                    :column-8 :diff
                                    :column-9 :%diff
                                    :column-10 :%port})
                (tc/select-rows (comp #(= % "Equities") :column-0))
                (tc/drop-columns (for [n [0 1 3 4 11 12 13 14 15]]
                                   (keyword (str "column-" n)))))

        ;; convert from string to numbers
        ds3 (-> ds2
                (tc/add-or-replace-columns {:marketvalue (remove-commas ds2 :marketvalue)
                                            :bookvalue (remove-commas ds2 :bookvalue)})
                (tc/convert-types {:qty :int16
                                   :marketvalue :float64
                                   :bookvalue :float64
                                   :diff :float64
                                   :%diff :float64
                                   :%port :float64}))

        ;; create the totals
        ds4 (-> ds3
                (tc/aggregate-columns [:qty :marketvalue :bookvalue :diff] #(reduce + %))
                (tc/add-columns {:%diff nil
                                 :%port nil
                                 :ticker "Totals"}))

        ;; concat the totals row to the rest of the dataset
        ds5 (tc/concat ds3 ds4)

        ;; format specific columns to currency (note using fn to save repetition)
        ds6 (-> ds5
                (tc/add-or-replace-columns (apply merge
                                                  (map (fn [k] {k (currency-format ds5 k)})
                                                       [:marketvalue :bookvalue])))
                #_(tc/add-or-replace-columns {:marketvalue (currency-format ds5 :marketvalue)
                                              :bookvalue (currency-format ds5 :bookvalue)})
                )
        ds7 (-> ds6
                (tc/dataset->str)
                (clojure.string/replace #".+\n\n" ""))
        ]
    (str acct
         "\n\n"
         ds7
         "\n\n\n\n")))





(defn -main
  "I don't do a whole lot ... yet."
  [fp]
  (let [ds (mk-ds fp)])
  ((comp pr-ds mk-ds) (first fpaths)))

(-main)





src/pradesigner/iepraderv2.clj
"
displays and summarizes results from ieprader
"

(ns pradesigner.ieprader
  (:require [tablecloth.api :as tc]
            [clojure.java.io :refer [file]])
  (:gen-class))


(def fpaths
  "defines the file paths for AccountHoldings"
  (let [base "/home/pradmin/dd/"
        fns (filter #(re-find #"^Account" %)
                    (seq (.list (file base))))]
    (map #(str base %) fns)))

(defn remove-commas
  "removes commas from numbers for conversion purposes"
  [ds col]
  (map #(clojure.string/replace % #"," "") (ds col)))

(defn currency-format
  "takes to 2 decimal places"
  [ds col]
  (map #(format "%.2f" %) (ds col)))

(defn mk-ds
  "makes dataset from filepath"
  [fp]
  (tc/dataset fp
              {:header-row? false
               :key-fn keyword}))

(defn get-acct
  "gets the account number from the dataset"
  [ds]
  (->> (first (ds :column-0))
       (re-find #"(^.+) -")
       (second)))

(defn prep-columns
  "remove extraneous columns and rename existing ones"
  [ds]
  (-> ds
      (tc/rename-columns {:column-2 :ticker
                          :column-5 :qty
                          :column-6 :marketvalue
                          :column-7 :bookvalue
                          :column-8 :diff
                          :column-9 :%diff
                          :column-10 :%port})
      (tc/select-rows (comp #(= % "Equities") :column-0))
      (tc/drop-columns (for [n [0 1 3 4 11 12 13 14 15]]
                         (keyword (str "column-" n))))))

(defn str->num
  "convert from string to numbers"
  [ds]
  (-> ds
      (tc/add-or-replace-columns {:marketvalue (remove-commas ds :marketvalue)
                                  :bookvalue (remove-commas ds :bookvalue)
                                  :diff (remove-commas ds :diff)})
      (tc/convert-types {:qty :int16
                         :marketvalue :float64
                         :bookvalue :float64
                         :diff :float64
                         :%diff :float64
                         :%port :float64})))

(defn concat-totals
  "create totals from columns :qty :marketvalue :bookvalue :diff and concat to ds"
  [ds]
  (tc/concat ds
             (-> ds
                 (tc/aggregate-columns [:qty :marketvalue :bookvalue :diff] #(reduce + %))
                 (tc/add-columns {:%diff nil
                                  :%port nil
                                  :ticker "Totals"}))))

(defn format->currency ;;TURNING BACK INTO STRING!!!!
  "format specific columns to currency (note using fn to save repetition)"
  [ds]
  (-> ds
      (tc/add-or-replace-columns (apply merge
                                        (map (fn [k] {k (currency-format ds k)})
                                             [:marketvalue :bookvalue :diff])))
      #_(tc/add-or-replace-columns {:marketvalue (currency-format ds5 :marketvalue)
                                    :bookvalue (currency-format ds5 :bookvalue)})))

(defn remove-ds-header
  "removes the ds header which shows filename and shape"
  [ds]
  (-> ds
      (tc/dataset->str)
      (clojure.string/replace #".+\n\n" "")))

(defn pr-ds
  "prepares ds given filepath"
  [fp]
  (let [ds (mk-ds fp)
        acct (get-acct ds)
        new-ds (-> ds
                   (prep-columns)
                   (tc/add-column :acct acct)
                   (tc/reorder-columns [:acct]))
        ;; ds-str (-> ds
        ;;            (prep-columns)
        ;;            (str->num)
        ;;            (concat-totals)
        ;;            (format->currency)
        ;;            (remove-ds-header))
        ]
;;    (str acct "\n\n" ds-str "\n\n\n\n")
    new-ds))

(defn mk-one-ds
  "concats all ds"
  [ds]
  (let [acct (get-acct ds)
        new-ds (-> ds
                   (prep-columns)
                   (str->num)
                   ;;(concat-totals)
                   ;;(format->currency)
                   (tc/add-column :acct acct)
                   (tc/reorder-columns [:acct]))]
    new-ds))

(defn join-all-ds
  "joins up all ds"
  [fp]
  (loop [acc (tc/dataset nil)
         fps fp]
    (if (empty? fps)
      acc
      (recur (tc/concat acc (mk-one-ds (mk-ds (first fps))))
             (rest fps)))))

(defn get-group-tots
  "produces a vector [total of totals, all group totals]"
  [ds group-by-cols cols-to-sum]
  (let [;;the summing function
        sumfn #(reduce + %)

        ;;the grouped ds
        g-ds (tc/group-by ds
                           group-by-cols
                           {:result-type :as-seq})

        ;;the group totals for each group
        group-sums (for [g g-ds]
                     (tc/aggregate-columns g
                                            cols-to-sum
                                            sumfn))

        ;;the group totals bound together into single dataset 
        summed-groups (apply tc/bind group-sums)

        ;;the total of the summed-groups dataset
        group-totals (tc/aggregate-columns summed-groups
                                            cols-to-sum
                                            sumfn)
        
        ;;the column sums attached to each group
        column-sums (map tc/bind g-ds group-sums)]

    ;;vector containing grouped dataset with
    ;;column sums and the column sums total
    [column-sums group-totals]
    ))


(get-group-tots (join-all-ds fpaths) [:acct] [:diff])
;; => [(null [9 8]:

|           :acct | :ticker | :qty | :marketvalue | :bookvalue |   :diff | :%diff | :%port |
|-----------------|---------|-----:|-------------:|-----------:|--------:|-------:|-------:|
| 58748842 prader |    TWOU |   20 |      1043.66 |     908.52 |  135.14 |  14.88 |   0.30 |
| 58748842 prader |     DDD |   40 |      1904.76 |    1879.01 |   25.75 |   1.37 |   0.55 |
| 58748842 prader |     MMM |   30 |      7292.79 |    7486.89 | -194.10 |  -2.59 |   2.12 |
| 58748842 prader |    DOCU |   20 |      7055.56 |    6853.15 |  202.41 |   2.95 |   2.05 |
| 58748842 prader |    VEGN |   30 |      1469.13 |    1414.78 |   54.35 |   3.84 |   0.43 |
| 58748842 prader |    SKLZ |  100 |      2748.00 |    2941.29 | -193.29 |  -6.57 |   0.80 |
| 58748842 prader |    TDOC |   40 |      8500.00 |    7425.69 | 1074.31 |  14.47 |   2.47 |
| 58748842 prader |    TSLA |   12 |     10130.35 |    9748.45 |  381.90 |   3.92 |   2.94 |
|                 |         |      |              |            | 1486.47 |        |        |
 null [4 8]:

|             :acct | :ticker | :qty | :marketvalue | :bookvalue |  :diff | :%diff | :%port |
|-------------------|---------|-----:|-------------:|-----------:|-------:|-------:|-------:|
| 59481160 lif_prad |       T |  200 |      7108.20 |    7074.75 |  33.45 |   0.47 |  10.69 |
| 59481160 lif_prad |     ENB |  200 |      9880.00 |    9534.95 | 345.05 |   3.62 |  14.85 |
| 59481160 lif_prad |     IBM |   44 |      7941.74 |    7988.89 | -47.15 |  -0.59 |  11.94 |
|                   |         |      |              |            | 331.35 |        |        |
 null [11 8]:

|                 :acct | :ticker | :qty | :marketvalue | :bookvalue |   :diff | :%diff | :%port |
|-----------------------|---------|-----:|-------------:|-----------:|--------:|-------:|-------:|
| 60359386 tfsa_ranjana |    ARKX |   40 |      1051.60 |    1021.27 |   30.33 |   2.97 |   1.18 |
| 60359386 tfsa_ranjana |    ARKQ |   10 |      1080.73 |    1010.73 |   70.00 |   6.93 |   1.21 |
| 60359386 tfsa_ranjana |    ARKW |   20 |      3867.06 |    3389.20 |  477.86 |  14.10 |   4.32 |
| 60359386 tfsa_ranjana |    ARKF |   20 |      1358.64 |    1232.30 |  126.34 |  10.25 |   1.52 |
| 60359386 tfsa_ranjana |    ARKG |   10 |      1153.89 |    1038.80 |  115.09 |  11.08 |   1.29 |
| 60359386 tfsa_ranjana |    ARKK |   30 |      4869.06 |    4134.07 |  734.99 |  17.78 |   5.44 |
| 60359386 tfsa_ranjana |      SQ |   30 |      9258.57 |    8836.50 |  422.07 |   4.78 |  10.35 |
| 60359386 tfsa_ranjana |    TRMB |   40 |      4103.68 |    3967.05 |  136.63 |   3.44 |   4.59 |
| 60359386 tfsa_ranjana |    PATH |   66 |      5639.11 |    5816.30 | -177.19 |  -3.05 |   6.30 |
| 60359386 tfsa_ranjana |       U |   33 |      4536.28 |    4559.57 |  -23.29 |  -0.51 |   5.07 |
|                       |         |      |              |            | 1912.83 |        |        |
 null [11 8]:

|               :acct | :ticker | :qty | :marketvalue | :bookvalue |   :diff | :%diff | :%port |
|---------------------|---------|-----:|-------------:|-----------:|--------:|-------:|-------:|
| 60359453 tfsa_kyron |    ARKX |   40 |      1051.60 |    1021.27 |   30.33 |   2.97 |   1.18 |
| 60359453 tfsa_kyron |    ARKQ |   10 |      1080.73 |    1009.63 |   71.10 |   7.04 |   1.21 |
| 60359453 tfsa_kyron |    ARKW |   20 |      3867.06 |    3386.75 |  480.31 |  14.18 |   4.34 |
| 60359453 tfsa_kyron |    ARKF |   20 |      1358.64 |    1232.06 |  126.58 |  10.27 |   1.52 |
| 60359453 tfsa_kyron |    ARKG |   10 |      1153.89 |    1038.18 |  115.71 |  11.15 |   1.29 |
| 60359453 tfsa_kyron |    ARKK |   30 |      4869.06 |    4135.53 |  733.53 |  17.74 |   5.46 |
| 60359453 tfsa_kyron |      SQ |   30 |      9258.57 |    8832.01 |  426.56 |   4.83 |  10.39 |
| 60359453 tfsa_kyron |    TRMB |   40 |      4103.68 |    3967.42 |  136.26 |   3.43 |   4.60 |
| 60359453 tfsa_kyron |    PATH |   66 |      5639.11 |    5816.30 | -177.19 |  -3.05 |   6.33 |
| 60359453 tfsa_kyron |       U |   33 |      4536.28 |    4563.37 |  -27.09 |  -0.59 |   5.09 |
|                     |         |      |              |            | 1916.10 |        |        |
 null [11 8]:

|              :acct | :ticker | :qty | :marketvalue | :bookvalue |   :diff | :%diff | :%port |
|--------------------|---------|-----:|-------------:|-----------:|--------:|-------:|-------:|
| 60359390 tfsa_prad |    ARKX |   40 |      1051.60 |    1021.27 |   30.33 |   2.97 |   1.17 |
| 60359390 tfsa_prad |    ARKQ |   10 |      1080.73 |    1009.87 |   70.86 |   7.02 |   1.20 |
| 60359390 tfsa_prad |    ARKW |   20 |      3867.06 |    3387.73 |  479.33 |  14.15 |   4.31 |
| 60359390 tfsa_prad |    ARKF |   20 |      1358.64 |    1231.81 |  126.83 |  10.30 |   1.51 |
| 60359390 tfsa_prad |    ARKG |   10 |      1153.89 |    1038.80 |  115.09 |  11.08 |   1.29 |
| 60359390 tfsa_prad |    ARKK |   30 |      4869.06 |    4135.90 |  733.16 |  17.73 |   5.43 |
| 60359390 tfsa_prad |      SQ |   30 |      9258.57 |    8828.52 |  430.05 |   4.87 |  10.32 |
| 60359390 tfsa_prad |    TRMB |   40 |      4103.68 |    3970.04 |  133.64 |   3.37 |   4.57 |
| 60359390 tfsa_prad |    PATH |   66 |      5639.11 |    5816.30 | -177.19 |  -3.05 |   6.28 |
| 60359390 tfsa_prad |       U |   33 |      4536.28 |    4561.12 |  -24.84 |  -0.55 |   5.06 |
|                    |         |      |              |            | 1917.26 |        |        |
) _unnamed [1 1]:

|   :diff |
|--------:|
| 7564.01 |
]



(defn do-printout
  "produces the output of the results"
  [[column-sums group-totals]]
  (let [colsums (for [cs column-sums]
                  (-> cs
                      tc/dataset->str
                      (clojure.string/replace #".+\n\n" "")))
        grtots (-> group-totals
                   tc/dataset->str
                   (clojure.string/replace #".+\n\n" ""))]
    (prn colsums "\n\n" grtots)))


(do-printout (get-group-tots (join-all-ds fpaths) [:acct] [:diff]))

(defn -main
  "runs get-group-tots"
  [& args]
  (prn (get-group-tots (join-all-ds fpaths) [:acct] [:diff])))





